[FLINK-18064][docs] Added unaligned checkpointing to docs. (#12722)
apache/flink: docs/ops/state/checkpoints.md

$ bin/flink run -s :checkpointMetaDataPath [:runArgs]
{% endhighlight %}

### Unaligned checkpoints

Starting with Flink 1.11, checkpoints can be unaligned (experimental). 
[Unaligned checkpoints]({% link concepts/stateful-stream-processing.md
%}#unaligned-checkpointing) contain in-flight data (i.e., data stored in
buffers) as part of the checkpoint state, which allows checkpoint barriers to
overtake these buffers. Thus, the checkpoint duration becomes independent of the
current throughput as checkpoint barriers are effectively not embedded into 
the stream of data anymore.

You should use unaligned checkpoints if your checkpointing durations are very

> high due to back-pressure. Then, checkpointing time becomes mostly

Reviewer: ```suggestion
high due to backpressure. Then, checkpointing time becomes mostly
```

Reviewer: Just for the sake of coherence.

independent of the end-to-end latency. Be aware unaligned checkpointing
adds to I/O to the state backends, so you shouldn't use it when the I/O to
the state backend is actually the bottleneck during checkpointing.

We flagged unaligned checkpoints as experimental as it currently has the

> following limitations:

Reviewer 2: I would also mention that `flatMap` operators can lead to unbounded spilled data.

- You cannot rescale from unaligned checkpoints. You have to take a savepoint 

> before rescaling. Savepoints are always aligned independent of the alignment

Author: Can you change the job graph with current checkpoints? I was always assuming that you need savepoints.

> before rescaling. Savepoints are always aligned independent of the alignment

Reviewer 3: Hmm...the current Flink docs [says](https://ci.apache.org/projects/flink/flink-docs-stable/ops/state/checkpoints.html#difference-to-savepoints) that retained checkpoints:
> do not support Flink specific features like rescaling.

...and nothing about the job graph.

Besides that, UC doesn't currently support Local recovery.

> before rescaling. Savepoints are always aligned independent of the alignment

Reviewer 4: Yes technically but its incidental. The community hasn't made any backward compat guaruntees around that behavior.

> before rescaling. Savepoints are always aligned independent of the alignment

Reviewer 3: I see, thanks. 
I think we can leave it as is then.

setting of checkpoints.
- Flink currently does not support concurrent unaligned checkpoints. However, 
due to the more predictable and shorter checkpointing times, concurrent 

> checkpoints might not be needed at all.

Reviewer 2: We should mention here that aligned savepoints also can not happen concurrently to unaligned checkpoint

> - Unaligned checkpoints may produce incorrect results for the following reasons.

Reviewer: ```suggestion
- Unaligned checkpoints may produce incorrect results for the following reasons:
```

Currently, Flink generates the watermark as a first step of recovery instead of 
storing the latest watermark in the operators to ease rescaling. In unaligned 
checkpoints, that means on recovery, **Flink generates watermarks after it 
restores in-flight data**. If your pipeline uses an **operator that applies the
latest watermark on each record**, it will produce **incorrect results** during 
recovery if the watermark is not directly or indirectly part of the operator 
state. Thus, **SQL OVER operator should not be used with unaligned

> checkpoints**, while window operators are safe to use. The work-around is to

Reviewer: ```suggestion
checkpoints**, while window operators are safe to use. The workaround is to
```

store the watermark in the operator state. If rescaling may occur, watermarks
should be stored per key-group in a union-state. We mostly likely will
implement this approach as a general solution (didn't make it into Flink 

> 1.11.0).

Reviewer 2: I think this paragraph is a bit too strong. As far as I understand, it's not that the UC will produce incorrect result, just that some records during the reprocessing might not be accounted as late data, right?

Author: I can tone it done, but basically we are breaking with the old assumption that watermarks don't need to be stored at the operator because they are sent first.
I'm especially referring to the OverITCases, which use a weird way to inject watermarks and logically should persist them. But now that I'm thinking about it, it's more a matter of the test setup itself.
